pkg/controller/staticpod/upgradeinfo/upgrade_info.go

```diff
@@ -128,64 +179,52 @@ func ReadyUpgradeWaitingNodes(infos map[string]*UpgradeInfo) []string {
 func ReadyNodes(infos map[string]*UpgradeInfo) []string {
 	var nodes []string
 	for node, info := range infos {
-		if info.Ready {
+		if info.NodeReady {
 			nodes = append(nodes, node)
 		}
 	}
 	return nodes
 }
 
-// UpgradeNeededNodes gets nodes that are not running the latest static pods
-func UpgradeNeededNodes(infos map[string]*UpgradeInfo) []string {
-	var nodes []string
+// ListOutUpgradeNeededNodesAndUpgradedNodes gets nodes that are not running the latest static pods and running the latest static pods
+func ListOutUpgradeNeededNodesAndUpgradedNodes(infos map[string]*UpgradeInfo) ([]string, []string) {
+	var upgradeNeededNodes, upgradeNodes []string
 	for node, info := range infos {
 		if info.UpgradeNeeded {
-			nodes = append(nodes, node)
-		}
-	}
-	return nodes
-}
-
-// UpgradedNodes gets nodes that are running the latest static pods
-func UpgradedNodes(infos map[string]*UpgradeInfo) []string {
-	var nodes []string
-	for node, info := range infos {
-		if !info.UpgradeNeeded {
-			nodes = append(nodes, node)
+			upgradeNeededNodes = append(upgradeNeededNodes, node)
+		} else {
+			upgradeNodes = append(upgradeNodes, node)
 		}
 	}
-	return nodes
+	return upgradeNeededNodes, upgradeNodes
 }
 
-// SetUpgradeNeededInfos sets `UpgradeNeeded` flag and counts the number of upgraded nodes
-func SetUpgradeNeededInfos(infos map[string]*UpgradeInfo, latestHash string) int32 {
-	var upgradedNumber int32
+// CalculateOperateInfoFromUpgradeInfoMap calculate the number of ready static pods, upgraded nodes,
+// the delete pods and whether all worker is finished.
+func CalculateOperateInfoFromUpgradeInfoMap(infos map[string]*UpgradeInfo) (int32, int32, bool, []*corev1.Pod) {
+	var (
+		upgradedNumber int32
+		readyNumber    int32
+		allSucceeded   = true
+		deletePods     = make([]*corev1.Pod, 0)
+	)
 
 	for _, info := range infos {
-		if info.StaticPod != nil {
-			if info.StaticPod.Annotations[StaticPodHashAnnotation] != latestHash {
-				// Indicate the static pod in this node needs to be upgraded
-				info.UpgradeNeeded = true
-				continue
-			}
+		// counts the number of ready static pods and upgraded nodes
+		if info.StaticPodReady {
+			readyNumber++
+		}
+		if !info.UpgradeNeeded {
 			upgradedNumber++
 		}
-	}
-
-	return upgradedNumber
-}
-
-// ReadyStaticPodsNumber counts the number of ready static pods
-func ReadyStaticPodsNumber(infos map[string]*UpgradeInfo) int32 {
-	var readyNumber int32
 
-	for _, info := range infos {
-		if info.StaticPod != nil {
-			if podutils.IsPodReady(info.StaticPod) {
-				readyNumber++
-			}
+		// sync worker pods info
+		if info.WorkerPodDeleteNeeded {
+			deletePods = append(deletePods, info.WorkerPod)
+		}
+		if info.WorkerPodStatusPhase != corev1.PodFailed && info.WorkerPodStatusPhase != corev1.PodSucceeded {
```

Comment:
Do we need to consider the existence of workerpod?
In this situation,`info.WorkerPodStatusPhase` may be "", and the result of `allSucceeded` is false

Reply:
yes, you are right.